### PR TITLE
Rewrote ``PlaybackParameterDialog``

### LINKED ISSUE #7960: Overhaul ``PlaybackParameterDialog``
I think the [PlaybackParameterDialog](https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/player/helper/PlaybackParameterDialog.java) needs an major overhaul/rework.
Many things are duplicated, could be done in a better way and the code is very confusing in some place. There is also a major lack of documentation/comments.

### PR DESCRIPTION
**Pull request by litetex**
#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [x] Meta improvement to the project (dev facing)

<details><summary>Backstory</summary>

* Noticed #7565 lately
* Took a look at changes inside there/``PlaybackParameterDialog`` <br/>
* This PR

</details>

#### Description of the changes in your PR
* Refactored the code
  * Removed a ton of duplicated stuff
  * Use viewbinding
  * Use ``Icepick`` instead of saving everything manually
* Tuned UI
  * UI fit's now - without a scroller - onto a 5in smartphone (Pixel 3a) in fullscreen mode
  * Removed "Adjust pitch by musical semitones" checkbox and replaced it by a tab like component that can be expanded on demand
  * Semitones now work with "hook"-mode
  * Mark the currently selected steps/pitchmode inside the UI
* Other
  * Used more containers in the UI so that we don't have to do everything inside the code

#### Before/After Screenshots/Screen Record
- Before:
grafik

- After:
grafik

https://user-images.githubusercontent.com/40789489/156843380-566584ad-61d7-4b17-bf97-d43317059b5d.mp4

#### Fixes the following issue(s)
- Fixes #7960

#### APK testing 
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

File: app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package org.schabi.newpipe.util
+
+import android.content.Context
+import android.graphics.drawable.Drawable
+import androidx.annotation.AttrRes
+
+/**
+ * Utility class for resolving [Drawables](Drawable)
+ */
+class DrawableResolver {
```
Comment (TacoTheDank) [inline app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt:10]: Regarding Kotlin, this should be an object class so that we don't need to use companion objects.

File: app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    companion object {
+        @JvmStatic
+        fun resolveDrawable(context: Context, @AttrRes attrResId: Int): Drawable? {
+            return androidx.core.content.ContextCompat.getDrawable(
+                context,
+                android.util.TypedValue().apply {
```
Comment (TacoTheDank) [inline app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt:16]: I think it would be better if these were replaced with imports instead of manually having them be written out.

File: app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt [hunk 3]
```diff
@@ -2,25 +2,24 @@ package org.schabi.newpipe.util
...
-                        this,
-                        true
-                    )
-                }.resourceId
-            )
-        }
+object DrawableResolver {
+    @JvmStatic
+    fun resolveDrawable(context: Context, @AttrRes attrResId: Int): Drawable? {
+        return androidx.core.content.ContextCompat.getDrawable(
```
Comment (TacoTheDank) [inline app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt:14]: Thanks for fixing! However, it seems you missed a spot :P

Comment (litetex) [reply app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt]: Thx for the hint.
I really shouldn't blindly copy past code from stackoverflow (or something like this) next time 😆

**Comment by karyogamy** [comment]
Nice work! By the way, would it be possible to collapse the 2 checkboxes into one line similar to `Percent` and `Semitone`? 
If the text is too long to fit, maybe we can just call them `Unhook` and `Skip Silence`? Even with text and subtext it might save more space.

**Comment by litetex** [comment]
@karyogamy 
I don't think that collapsing the checkboxes would be beneficial, because we don't need to save more space (already optimized for a standard 5" screen).
I might eventually do more UI optimizations in a future PR (e.g. representing the "Unhook" checkbox as a chain between speed and pitch) but the current one primarily aims at code improvements.

File: app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import android.content.Context
+import android.graphics.drawable.Drawable
+import android.util.TypedValue
+import androidx.annotation.AttrRes
+import androidx.core.content.ContextCompat
+
+/**
+ * Utility class for resolving [Drawables](Drawable)
+ */
+object DrawableResolver {
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt:12]: Why is this here and not in `ThemeHelper`? I'd prefer not to create yet another class in the `util` package when there is already one that contains similar functions

File: app/src/main/java/org/schabi/newpipe/player/helper/PlayerSemitoneHelper.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Converts between percent and 12-tone equal temperament semitones.
+ * <br/>
+ * @see
+ * <a href="https://en.wikipedia.org/wiki/Equal_temperament#Twelve-tone_equal_temperament">
+ *     Wikipedia: Equal temperament#Twelve-tone equal temperament
+ * </a>
+ */
+public final class PlayerSemitoneHelper {
+    public static final int TONES = 12;
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/helper/PlayerSemitoneHelper.java:12]: I'd rename this to `SEMITONE_COUNT` or something like that, since `TONES` is different from `SEMITONES` (a tone is twice a semitone)

File: app/src/main/java/org/schabi/newpipe/player/helper/PlaybackParameterDialog.java [hunk 6]
```diff
@@ -1,122 +1,120 @@
...
+import icepick.State;
+
 public class PlaybackParameterDialog extends DialogFragment {
+    private static final String TAG = "PlaybackParameterDialog";
+
     // Minimum allowable range in ExoPlayer
-    private static final double MINIMUM_PLAYBACK_VALUE = 0.10f;
-    private static final double MAXIMUM_PLAYBACK_VALUE = 3.00f;
+    private static final double MIN_PLAYBACK_VALUE = 0.10f;
+    private static final double MAX_PLAYBACK_VALUE = 3.00f;
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/helper/PlaybackParameterDialog.java:48]:
```suggestion
    private static final double MIN_PLAYBACK_SPEED = 0.10f;
    private static final double MAX_PLAYBACK_SPEED = 3.00f;
```

**Review by Stypox — CHANGES_REQUESTED** [review]
> Used more containers in the UI so that we don't have to do everything inside the code

This is not good for performance afaik, but if we don't migrate to `ConstraintLayout` (which has view groups) there is no alternative

File: app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt (hunk 4 above)
Comment (litetex) [reply app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt:12]: ``ThemeHelper`` is used for doing theme-related stuff (e.g. ``getThemeForService``). ``DrawableResolver`` is used for resolving Drawables.

Not sure how these classes should be related...

Comment (Stypox) [reply app/src/main/java/org/schabi/newpipe/util/DrawableResolver.kt:12]: No, `ThemeHelper` contains things that have to do with theme and Context resources. Those two things are tightly coupled since the resources are determined by the theme. Hence why, for example, `resolveColorFromAttr` is in that class, and `resolveDrawable` should go there too. Before #5927 `resolveResourceIdFromAttr` existed in that class, too.

**Comment by litetex** [comment]
> > Used more containers in the UI so that we don't have to do everything inside the code
>
> This is not good for performance afaik, but if we don't migrate to `ConstraintLayout` (which has view groups) there is no alternative (so nothing to do)

1. It aren't so many new "containers" (1-3)
2. I think that `ConstraintLayout` is overrated in the first place - at least for this view:

So I quickly googled why `Constraintlayout` is better than `Relativelayout` and found these sources (from https://stackoverflow.com/questions/59628341/constraintlayout-vs-relativelayout-performance-2020):
* https://developer.android.com/topic/performance/rendering/optimizing-view-hierarchies.html#cheaper
  > In the N release of Android, the ConstraintLayout class provides similar functionality to RelativeLayout, but at a significantly lower cost.
  
  However it doesn't cite any sources why it performs better...
* https://android-developers.googleblog.com/2017/08/understanding-performance-benefits-of.html
  This article is in my opinion not representable:
  * It simply removes all relative layouts when using `ConstraintLayout`. They don't use any view groups.
  * According to their graph the performance in the "Measurement and Layout phase" increased from 0.65ms to 0.5ms (which is a performance increase in 25-30% not 40% as they stated)

TL;DR
I will not rewrite the complete layout now to ``ConstraintLayout`` because it's then 0.15ms faster. Especially regarding that the dialog/layout is rarely used in the first place.
We can create an issue that does this in the follow up, but for now I think that the current state should be enough.

**Comment by Stypox** [comment]
> I will not rewrite the complete layout now to ConstraintLayout because it's then 0.15ms faster. Especially regarding that the dialog/layout is rarely used in the first place.
We can create an issue that does this in the follow up, but for now I think that the current state should be enough.

Yes, I do agree it shouldn't be done here, and I do agree there is no big need to spend time on it. It's just something worth noting in my opinion, since `ConstraintLayout`s are better not only in performance but also they are also simpler to code, cause less errors and are more powerful.

**Comment by SameenAhnaf** [comment]
Do we really need to click twice just to save space? A checkbox could be placed here instead? 
IMG_20220403_173456

**Comment by litetex** [comment]
@SameenAhnaf 
1. How often do you change between semitones and percent?Not that often or?
2. If the "pitch" and "Semtiones" are displayed on a small screen and have long localizations they won't fit

**Comment by Redirion** [comment]
I resolved the conflict directly here on github and it resulted in that not really helpful commit message "Merge branch 'dev'". Can you clean that up please litetex? This should be ready to merge.

**Review by Stypox — APPROVED** [review]
I tested and it works well. Thank you!

**Comment by yashpalgoyal1304** [comment]
> * Mark the currently selected steps/pitchmode inside the UI

hey! is this reverted in v0.24.1 ?? I was unable to see the rectangular dashed border around currently selected speed selector/speed adjustment step size buttons in that version.

----
<details>
<summary>Trail/please ignore</summary>

* blame on [dialog_playback_parameter.xml](https://github.com/TeamNewPipe/NewPipe/blame/dev/app/src/main/res/layout/dialog_playback_parameter.xml)
* <-- files in [!9145](https://github.com/TeamNewPipe/NewPipe/pull/9145/files#diff-9e225b66a55f2a93310079e15a6616443352f4008509b9db1e65b0bd821a0224)
* <-- linked in [#8962](https://github.com/TeamNewPipe/NewPipe/issues/8962#ref-pullrequest-1412514715)
* <-- gh issue 🔍 : [adjustment step size](https://github.com/TeamNewPipe/NewPipe/issues?q=adjustment+step+size)
* <-- search term found in [!7728](https://github.com/TeamNewPipe/NewPipe/pull/7728)
* <-- Release 0.23.0 [#8230](https://github.com/TeamNewPipe/NewPipe/issues/8230)
* <-- gh issue 🔍 : [speed selector](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+speed+selector+is%3Aclosed)

</details>

**Comment by yashpalgoyal1304** [comment]
> I can see it just fine. Maybe your OS colours are different?

i just rechecked, and yeah, it was present there.... though it was very faint due to low resolution and small screen size of the device i saw that on.
